### ISSUE TITLE
Improving performance of ApsimClassic in Singularity Container

### ISSUE THREAD

**Issue by youngr1960** [APSIMInitiative/APSIMClassic#2094]
I have created a Singularity container for ApsimClassic based on Debian and am trying to improve the performance of Apsim within the container. Details of the container are:
Singularity version:   3.7.0
Debian version:        11.0
Mono version:           6.12
Apsim version:           7.10-280921
HPC OS version:         RHEL 8.2

The configuration of the container is based on older docker and mono container. During the build process there doesn't seem to be any errors and when Apsim is run there are no errors in the output files. From this I am assuming Apsim is running correctly. Based on user feedback, they are saying Apsim in the container is running slow compared to running on a Windows computer, e.g. 700s compared to 70s. I have been trying version options with the container, mono and Apsim to improve performance. I was just wondering has anybody got options or hints to improve the runtime of Apsim jobs. Thank you

**Comment by BrianCollinss** [comment]
Have you made sure you assigned enough CPU cores to the container?

**Comment by youngr1960** [comment]
Generally speaking, users don't simply start the container and run ApsimClassic within the container. As we are using a HPC each job has in its command file the following:

export NUMBER_OF_PROCESSORS=128
singularity exec /opt/sw/singularity/images/ApsimClassicv2.sif ApsimRun.sh ${apsiminputfile}

So each job is allocated 128 cores, which are all used when a compute node running Apsim is monitored. Also, in my test runs I have allocated 256Gb of RAM to each job.

**Comment by peter-devoil** [comment]
Slow throughput must come from either cpu or disk bottlenecks - you've identified that cpus are being used, but I wonder if they running apsim or fighting over a single nfs share? Depending on how the user has configured apsim, they may be doing lots of small IO writes that the kernel has to service somehow. Running on a scratch partition may improve things

My experiences: 
- RAM generally isnt an issue if you allocate 1Gb / CPU

- I've stuck with mono 4.8.1 as I saw spikes in resource usage with mono 6, which the hpc scheduler didn't like and subsequently killed.

- Another quirk I have found with singularity is that it randomly fails when you start a lot of concurrent instances (eg one per CPU), I don't think you're doing that here.

**Comment by youngr1960** [comment]
When the hpc was first configured, yes there was problems with IO to nfs shares, so the /sandisk1 was reconfigured to provided improved IO which provided about 30% in speed improvement. To improve performance more I am now converting scripts to use /localscratch on the local harddrive which should technically improve performance more than using an nfs share. Yes, when apsim was first installed it was getting killed by the scheduler because it was grabbing all compute node cpu resources, however, I haven't seen the scheduler kill a job when apsim is running via singularity and mono 6.

**Comment by youngr1960** [comment]
Further to my last comments, I have managed to reduce a test job from 80 to about 12 minutes by using a combination of optimised nfs share and local scratch disk. One oddity I have noticed, is the test jobs seem to run quicker with 64 cores than being run on 128 cores. There doesn't seem to be any errors on compute nodes, e.g. io/inodes, that would cause this oddity. The only thing I can think of is the management of files being created, but I can't prove it. Has anybody seen this oddity and possibly found a way around it. Thanks
